Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm

```diff
@@ -1333,6 +1485,102 @@ - (void)_applicationDidBecomeActive:(NSNotification*)notification
+    adjustedFullscreenWindowFrame.origin.y = (sceneSize.height  - CGRectGetHeight(adjustedFullscreenWindowFrame)) / 2;
+
+    // Workaround rdar://103936581. Since `originalWindow` is not owned by WebKit, it is currently
+    // not possible to prevent it from participating in scene resize, as it is being faded out.
+    // Temporarily change the object's class to override scene resize behavior.
+    object_setClass(_lastKnownParentWindow.get(), [WKFixedSizeWindow class]);
+    object_setClass(_window.get(), [WKFixedSizeWindow class]);
+
+    UIWindowScene *scene = [_lastKnownParentWindow windowScene] ?: [_window windowScene];
+
+    scene.mrui_placement.preferredChromeOptions = RSSSceneChromeOptionsNone;
+
+    WebKit::resizeScene(scene, sceneSize, [strongSelf = retainPtr(self), self, adjustedOriginalWindowFrame, adjustedFullscreenWindowFrame]() {
+        [_lastKnownParentWindow setFrame:adjustedOriginalWindowFrame];
+        [_window setFrame:adjustedFullscreenWindowFrame];
+    });
+}
+
+- (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(CompletionHandler<void()>&&)completionHandler
+{
+    WKFullScreenWindowController *controller = self;
+    UIWindow *inWindow = enter ? _window.get() : _lastKnownParentWindow.get();
+    UIWindow *outWindow = enter ? _lastKnownParentWindow.get() : _window.get();
+    WKFullScreenParentWindowState *originalState = _parentWindowState.get();
+
+    inWindow.transform3D = CATransform3DTranslate(originalState.transform3D, 0, 0, kIncomingWindowZOffset);
+
+    if ([[NSUserDefaults standardUserDefaults] boolForKey:@"WebKitEnableFullScreenDimming"]) {
+        [UIView animateWithDuration:kDarknessAnimationDuration animations:^{
+            MRUIStage *stage = UIApplication.sharedApplication.mrui_activeStage;
+            stage.preferredDarkness = enter ? MRUIDarknessPreferenceVeryDark : originalState.preferredDarkness;
+        } completion:nil];
+    }
+
+    [UIView animateWithDuration:kOutgoingWindowFadeDuration delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
+        outWindow.alpha = 0;
+    } completion:nil];
+
+    [UIView animateWithDuration:kOutgoingWindowTranslationDuration delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
+        outWindow.transform3D = CATransform3DTranslate(outWindow.transform3D, 0, 0, kOutgoingWindowZOffset);
+    } completion:nil];
+
+    [UIView animateWithDuration:kIncomingWindowTranslationDuration delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
+        inWindow.transform3D = originalState.transform3D;
+    } completion:nil];
+
+    auto completion = makeBlockPtr([controller = retainPtr(controller), inWindow = retainPtr(inWindow), originalState = retainPtr(originalState), enter, completionHandler = WTFMove(completionHandler)] (BOOL finished) mutable {
+        if (!finished)
+            return;
```

Comment:
(Just curious — is it intended to skip the `completionHandler` in the case of `!finished`?)

Reply:
Not intentional... I should remove this check as the animation isn't intended to be interruptible. Probably better in a follow-up, to avoid mixing with upstreaming.

Reply:
Yep! Definitely good to keep this upstreaming patch as-is.